File: app/src/main/java/me/li2/android/architecture/utils/DateUtils.java
package me.li2.android.architecture.utils;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;

/**
 * Created by weiyi on 29/7/18.
 * https://github.com/li2
 */
public class DateUtils {

    private static final String TAG = DateUtils.class.getSimpleName();

    private static final String UTC_FULL_TIME_PATTERN ="yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";

    public static Date getDateWithPattern(@NonNull String timestamp, @NonNull String pattern, @Nullable TimeZone timeZone) {
        Date result = null;
        try {
            SimpleDateFormat sdf = new SimpleDateFormat(pattern);
            if (timeZone != null) {
                sdf.setTimeZone(timeZone);
            }
            result = sdf.parse(timestamp);
        } catch (ParseException | IllegalArgumentException e) {
            Log.e(TAG, "Error getDateWithPattern(" + timestamp + ")" + e.getMessage());
        }
        return result;
    }

    public static Date getUtcDate(@NonNull String timestamp) {
        return getDateWithPattern(timestamp, UTC_FULL_TIME_PATTERN, TimeZone.getTimeZone("UTC"));
    }

    public static long getDaysBetween(Date startDate, Date endDate) {
        final long diff = endDate.getTime() - startDate.getTime();
        if (diff <= 0) {
            return 0;
        }
        long days = diff / TimeUnit.DAYS.toMillis(1);
        return days;
    }
}

File: app/src/main/java/me/li2/android/architecture/utils/DateUtils.kt
package me.li2.android.architecture.utils

import java.text.SimpleDateFormat
import java.util.*

const val DATE_PATTERN_yyyy_MM_dd_T_HH_mm_ss = "yyyy-MM-dd'T'HH:mm:ss"
const val DATE_PATTERN_dd_MM_yyyy = "dd/MM/yyyy"

fun getLocalDateFromString(timestamp: String?, pattern: String): Date? =
        if (timestamp != null) getDateFromStringWithPattern(timestamp, pattern, TimeZone.getDefault()) else null

internal fun getDateFromStringWithPattern(timestamp: String, pattern: String, timeZone: TimeZone?): Date? =
        try {
            val sdf = SimpleDateFormat(pattern)
            timeZone?.let {
                sdf.timeZone = timeZone
            }
            sdf.parse(timestamp)
        } catch (e: Exception) {
            println("Error getDateFromStringWithPattern($timestamp, $pattern): $e.message")
            null
        }
